rasa/core/events/__init__.py

```diff
@@ -1010,11 +1011,27 @@ def __str__(self) -> Text:
     def __hash__(self) -> int:
         return hash(self.action_name)
 
+    def _check_for_e2e_text(self) -> bool:
+        if hasattr(self, "e2e_text"):
+            return self.e2e_text is not None
+        else:
+            return False
+
     def __eq__(self, other) -> bool:
         if not isinstance(other, ActionExecuted):
             return False
         else:
-            return self.action_name == other.action_name
+            # if both have e2e_text compare both action name and e2e text
+            if hasattr(self, "e2e_text") and hasattr(other, "e2e_text"):
+                return (
+                    self.action_name == other.action_name
+                    and self.e2e_text == other.e2e_text
+                )
+            # if both of them don't have the attribute, only compare action name
+            if self._check_for_e2e_text() == other._check_for_e2e_text():
+                return self.action_name == other.action_name
+            else:
+                return False
```

Comment:
why do we need this?

Reply:
I've simplified it a bit; in essence now when there is `e2e_text` in both `ActionExecuted`'s, we need to check it is equal to say that the `ActionExecuted`'s are equal

Reply:
where do we need the comparison?

Reply:
off the top of my head, in many tests; but I cant guarantee its not used anywhere else

Reply:
If it's just for the tests then we shouldn't add it to the production code.

Reply:
Do you mean that the `__eq__` function should not be tehre at all?

Reply:
In general I'd say: don't add production code which isn't required in production (e.g. only for tests). So I'd simply revert your changes. As the whole `__eq__` is probably only used in tests it's weird anyway :-D Imo it's also ok to keep it then. I'd just remove the `hasattr` then as it will always have the attribute